sktime/forecasting/compose/_pipeline.py

```diff
@@ -1358,21 +1382,31 @@ def _fit(self, y, X, fh):
         # remember if X seen was None
         self.X_was_None_ = X is None
 
-        if self.behaviour == "update" and X is not None:
+        # initialize forecaster_X_ and forecaster_y_
+        self.forecaster_y_ = self.forecaster_y.clone()
+        if X is not None:
             self.forecaster_X_ = self.forecaster_X_c.clone()
-            self.forecaster_X_.fit(y=self._get_Xcols(X), fh=fh_X)
 
-        self.forecaster_y_ = self.forecaster_y.clone()
-        if self.fit_behaviour == "use_actual":
-            self.forecaster_y_.fit(y=y, X=X, fh=fh)
+        if self.behaviour == "update" and X is not None:
+            X_for_fcX = self._get_X_for_fcX(X)
+            self.forecaster_X_.fit(y=self._get_Xcols(X), fh=fh_X, X=X_for_fcX)
+
+        if X is None or self.fit_behaviour == "use_actual":
+            X_for_fcy = X
         elif self.fit_behaviour == "use_forecast":
             if not self.forecaster_X_.get_tag("capability:insample"):
                 raise ValueError(
                     "forecaster_X does not have `capability:insample`. "
                     "Thus, it is not valid with `fit_behaviour=use_forecast`."
                 )
-            x_insample = self.forecaster_X_.predict(fh=X.index, X=X)
-            self.forecaster_y_.fit(y=y, X=x_insample, fh=fh)
+            if isinstance(X.index, pd.MultiIndex):
+                X_times = X.index.get_level_values(-1).unique()
+            else:
+                X_times = X.index
+            fh_for_fcst = ForecastingHorizon(X_times, is_relative=False)
```

Comment:
of course. As the PR header mentions, an error would occur in the case of integer index. This is because naively passing `X.index` coerces it to `ForecastingHorizon`, which assumes relative in case of integers, e.g., 1 is 1 step ahead of the latest time stamp seen. But we need absolute (as we want to predict in sample).

This was not covered by tests and would simply result in an exception, as we later assume that the indices of the forecast are identical to `X.index` - which in the above failure condition they are not (i.e., we forecast 1 step ahead instead of absolute time stamp 1 etc).

Reply:
To tell `sktime` that we want absolute, we need to construct the `ForecastingHorizon` explicitly.

The error above does not occur with datetime-s, and that's probably where @benHeid tested it (and possibly the reviewer back then who might have been me).